File: source_formatting/dibur_hamatchil.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

_TITLES_TO_BOLDIFY = (
    "Rashi",
    "Tosafot",
)

_SPLITTERS = (
    " - ",
    " – ",
)

def bold_diburei_hamatchil(text, commentary_name):
    if commentary_name not in _TITLES_TO_BOLDIFY:
        return text

    for splitter in _SPLITTERS:
        if splitter in text:
            dibur_hamatchil, comment = text.split(splitter, 1)
            return f'<strong class="dibur-hamatchil">{dibur_hamatchil}</strong>{splitter}{comment}'

    return text

File: source_formatting/dibur_hamatchil.ts
const TITLES_TO_BOLDIFY = new Set([
  "Ran",
  "Rashi",
  "Rashbam",
  "Tosafot",
  "Mishnah Berurah",
]);

const SPLITTERS = [
  "-",
  "–",
].map(x => ` ${x} `);

function format(diburHamatchil: string, splitter: string, comment: string) {
  return `<strong class="dibur-hamatchil">${diburHamatchil}</strong>${splitter}${comment}`;
}

export function boldDibureiHamatchil(
  text: sefaria.TextType,
  commentaryName: string,
): sefaria.TextType {
  if (!TITLES_TO_BOLDIFY.has(commentaryName)) {
    return text;
  }

  if (Array.isArray(text)) {
    return text.map(x => boldDibureiHamatchil(x, commentaryName)) as sefaria.TextType;
  }

  for (const splitter of SPLITTERS) {
    const index = text.indexOf(splitter);
    if (index !== -1) {
      const diburHamatchil = text.substring(0, index);
      const comment = text.substring(index + splitter.length);
      return format(diburHamatchil, splitter, comment);
    }
  }

  if (commentaryName === "Tosafot" && text.includes(". ")) {
    const firstPeriod = text.indexOf(". ");
    const diburHamatchil = text.substring(0, firstPeriod);
    // Only match a period it's in the beginning of the text. As a simple heuristic, use the first
    // 20 words of the comment.
    if (diburHamatchil.split(" ").length <= 20) {
      return format(diburHamatchil, " - ", text.substring(firstPeriod + 2));
    }
  }

  return text;
}
